### PR TITLE
Small parser improvements

### LINKED ISSUE #1227: Removing u""
After looking through #1226, I realized the `u` prefix is discarded. Why is this even allowed? Seems like it'd be potential cause for confusion for a Hy newbie. IMO, it would be better if it printed an error saying that all Hy strings are already Unicode.

Thoughts?

### LINKED ISSUE #2239: In code with DOS-style newlines, literal newlines are treated differently in double-quoted strings and bracket strings
This bug is most relevant to Windows users, but can be observed on Linux. For example, make a file with the contents

```
(print (repr r"hello
goodbye"))
(print (repr #[[hello
goodbye]]))
(print (repr #[[
hello
goodbye]]))
```

where the newlines are all 0x0D 0x0A (`unix2dos` is handy). The output is then

```
'hello\ngoodbye'
'hello\r\ngoodbye'
'\r\nhello\r\ngoodbye'
```

So in the case of double-quoted strings, Hy reads in `"\n"` for the newlines regardless of the newline style in the source code, as Python does for triple-quoted strings, but bracket strings keep whatever newline style was used in the source. (Furthermore, Hy doesn't trim a leading newline, as it's supposed to for bracket strings, if that newline is 0x0D 0x0A.) My feeling is that Python made the wrong choice here with this bit of magic, particularly for raw strings, but we should probably aim for consistency with it.

Tests for any fix to this bug should probably add the step `run: git config --global core.autocrlf false` to `.github/workflows/tests.yml`. By default, the GitHub Actions runner for Windows translates all newlines at checkout-time, which can complicate things.

### LINKED ISSUE #2254: Invalid string escape sequences become literal backslashes
Python's traditional behavior given something like `"a\p"` is to silently treat it as `"a\\p"` (or equivalently, `r"a\p"`), because there is no escape sequence `\p`. Recent Pythons will produce a `DeprecationWarning` for this, although you may have to opt in to the warning. Hy currently treats  `"a\p"` as `"a\\p"` with no warning and no error.

Since we're still not at 1.0, I think we might as well jump headfirst into the future and treat invalid escape sequences as syntax errors, regardless of the underlying Python version. The literal-backslash fallback isn't useful and is liable to get the programmer into trouble.

### PR DESCRIPTION
**Pull request by Kodiologist**
- Closes #1227.
- Closes #2239.
- Closes #2254.

### CODE REVIEW

**Comment by allison-casey** [comment]
Looks good from my cursory glance. I'd only ask that you link to the Python page about unrecognized escape sequences being deprecated/syntax errors. I'll try and give this a more thorough review today/tomorrow

**Comment by Kodiologist** [comment]
Sure, done.

File: hy/lex/hy_reader.py [hunk 1]
```diff
@@ -249,6 +249,12 @@ def keyword(self, _):
 
     @reader_for('"')
     def prefixed_string(self, _, prefix=""):
+        prefix_chars = set(prefix)
+        if (len(prefix_chars) != len(prefix) or prefix_chars - set("bfr")) and not (
+            set("bf") <= prefix_chars
+        ):
+            raise LexException.from_reader(f"invalid string prefix {prefix!r}", self)
+
         escaping = False
```
[the remark this answers is not in the document]
Comment (Kodiologist) [reply hy/lex/hy_reader.py:258]: `set("bf") <= prefix_chars` is supposed to catch that. Are you sure it doesn't?

[the remark this answers is not in the document]
Comment (Kodiologist) [reply hy/lex/hy_reader.py:260]: It was a simple logic bug. Whoopsies. It should be fixed now. Thanks.

[the remark this answers is not in the document]
Comment (allison-casey) [reply hy/lex/hy_reader.py:260]: :partying_face:

**Review by allison-casey — APPROVED** [review]

**Review by scauligi — APPROVED** [review]

**Pull request merged**